### PR TITLE
iptables: ugly hack to quickly fix kubernetes on RHEL 7

### LINKED ISSUE #82587: kube-proxy fails to restore iptable rules
*What happened**:
Running `hack/local-up-cluster.sh`, kube-proxy never succeeds calling `iptables-restore --noflush --counters`:

```
I0911 14:30:08.104508    1022 proxier.go:1420] Restoring iptables rules: *filter
:KUBE-SERVICES - [0:0]
:KUBE-EXTERNAL-SERVICES - [0:0]
:KUBE-FORWARD - [0:0]
-A KUBE-FORWARD -m conntrack --ctstate INVALID -j DROP
-A KUBE-FORWARD -m comment --comment "kubernetes forwarding rules" -m mark --mark 0x00004000/0x00004000 -j ACCEPT
COMMIT
*nat
:KUBE-SERVICES - [0:0]
:KUBE-NODEPORTS - [0:0]
:KUBE-POSTROUTING - [0:0]
:KUBE-MARK-MASQ - [0:0]
:KUBE-SVC-NPX46M4PTMTKRN6Y - [0:0]
:KUBE-SEP-SGYQAO2MFIR4HIEV - [0:0]
:KUBE-SVC-TCOU7JCQXEZGVUNU - [0:0]
:KUBE-SEP-SNPTLXDNVSPZ5ND2 - [0:0]
:KUBE-SVC-ERIFXISQEP7F7OF4 - [0:0]
:KUBE-SEP-7PPXA5JT5ALVQPIV - [0:0]
-A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -m mark --mark 0x00004000/0x00004000 -j MASQUERADE
-A KUBE-MARK-MASQ -j MARK --set-xmark 0x00004000/0x00004000
-A KUBE-SERVICES -m comment --comment "default/kubernetes:https cluster IP" -m tcp -p tcp -d 10.0.0.1/32 --dport 443 -j KUBE-SVC-NPX46M4PTMTKRN6Y
-A KUBE-SVC-NPX46M4PTMTKRN6Y -j KUBE-SEP-SGYQAO2MFIR4HIEV
-A KUBE-SEP-SGYQAO2MFIR4HIEV -s 192.168.122.150/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-SGYQAO2MFIR4HIEV -m tcp -p tcp -j DNAT --to-destination 192.168.122.150:6443
-A KUBE-SERVICES -m comment --comment "kube-system/kube-dns:dns cluster IP" -m udp -p udp -d 10.0.0.10/32 --dport 53 -j KUBE-SVC-TCOU7JCQXEZGVUNU
-A KUBE-SVC-TCOU7JCQXEZGVUNU -j KUBE-SEP-SNPTLXDNVSPZ5ND2
-A KUBE-SEP-SNPTLXDNVSPZ5ND2 -s 172.17.0.2/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-SNPTLXDNVSPZ5ND2 -m udp -p udp -j DNAT --to-destination 172.17.0.2:53
-A KUBE-SERVICES -m comment --comment "kube-system/kube-dns:dns-tcp cluster IP" -m tcp -p tcp -d 10.0.0.10/32 --dport 53 -j KUBE-SVC-ERIFXISQEP7F7OF4
-A KUBE-SVC-ERIFXISQEP7F7OF4 -j KUBE-SEP-7PPXA5JT5ALVQPIV
-A KUBE-SEP-7PPXA5JT5ALVQPIV -s 172.17.0.2/32 -j KUBE-MARK-MASQ
-A KUBE-SEP-7PPXA5JT5ALVQPIV -m tcp -p tcp -j DNAT --to-destination 172.17.0.2:53
-A KUBE-SERVICES -m comment --comment "kubernetes service nodeports; NOTE: this must be the last rule in this chain" -m addrtype --dst-type LOCAL -j KUBE-NODEPORTS
COMMIT
I0911 14:30:08.104638    1022 iptables.go:397] running iptables-restore [--noflush --counters]
E0911 14:30:08.106159    1022 proxier.go:1423] Failed to execute iptables-restore: exit status 4 (Another app is currently holding the xtables lock. Perhaps you want to use the -w option?
)
```

No matter how long I wait, kube-proxy always gets `Another app is currently holding the xtables lock`.

Adding `-w` to iptables-restore cmdline (via hacked version check) fixed the issue for me, but I am not able to judge what else it break (there must have been reason for https://github.com/kubernetes/kubernetes/pull/80368). Until now I was happy user of `local-up-cluster.sh` and I'd like have it working again.

**Anything else we need to know?**:

kernel-3.10.0-957.12.2.el7.x86_64
iptables-1.4.21-28.el7.x86_64
kubernetes v1.16.0-rc.1

/sig network

@danwinship, PTAL
cc @bertinatto

### PR DESCRIPTION
**Pull request by danwinship**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
I broke kubernetes on RHEL 7. This is not the right fix, but it's the most minimal fix.

**Which issue(s) this PR fixes**:
Addresses #82587 but is not the final fix

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```
(Or do we release-note differences between pre-release versions?)

/sig network
/priority critical-urgent

### CODE REVIEW

**Comment by danwinship** [comment]
/hold
still testing

File: pkg/util/iptables/iptables.go [hunk 1]
```diff
@@ -581,6 +581,8 @@ func getIPTablesWaitFlag(version *utilversion.Version) []string {
 func getIPTablesRestoreWaitFlag(version *utilversion.Version) []string {
 	if version.AtLeast(WaitRestoreMinVersion) {
 		return []string{WaitString, WaitSecondsValue}
+	} else if version.String() == "1.4.21" { // HACK for 1.16.0
```
Comment (neolit123) [inline pkg/util/iptables/iptables.go]: if this is an ugly hack that we don't want around, this comment should link to https://github.com/kubernetes/kubernetes/issues/82587
and we should keep that issue open.

**Comment by danwinship** [comment]
/hold cancel
confirmed that this fixes the bug

**Comment by danwinship** [comment]
node-e2e seems to be flaking for everyone

**Comment by liggitt** [comment]
was supposed to be fixed in https://github.com/kubernetes/test-infra/pull/14291, though the job that failed says it was on that commit of test-infra

/retest

**Comment by liggitt** [comment]
oh, that job failed with a different (spurious?) error:

> I0911 20:53:40.226] Call:  gsutil -q -h Content-Type:application/json cp /tmp/gsutil_VtazFi gs://kubernetes-jenkins/pr-logs/pull/82596/pull-kubernetes-node-e2e/1171889251453767680/started.json
> W0911 21:07:40.523] ServiceException: 500 Backend Error

File: pkg/util/iptables/iptables.go [hunk 2]
```diff
@@ -581,6 +581,8 @@ func getIPTablesWaitFlag(version *utilversion.Version) []string {
 func getIPTablesRestoreWaitFlag(version *utilversion.Version) []string {
 	if version.AtLeast(WaitRestoreMinVersion) {
 		return []string{WaitString, WaitSecondsValue}
+	} else if version.String() == "1.4.21" { // HACK for 1.16.0; see https://github.com/kubernetes/kubernetes/issues/82587
```
Comment (thockin) [inline pkg/util/iptables/iptables.go]: Is there ANY other way to cross check this against something else to filter out false-positives?

**Review by thockin — COMMENTED** [review]
I'd really rather not rollback the original changes at this point, so I am inclined to let this fly, but it is a little worrisome.

**Comment by mrunalp** [comment]
We are testing this on CRI-O CI and it seems to be working so far. (Our clusters were failing to come up but with this PR they are.)

File: pkg/util/iptables/iptables.go (hunk 2 above)
Comment (danwinship) [reply pkg/util/iptables/iptables.go]: I was trying to minimize the amount of new code. The right way to test would be to bring back the original code that ran `iptables-restore --version` to see how it behaves. That would look like https://github.com/danwinship/kubernetes/commit/75888077, which I've confirmed passes unit tests but haven't tested beyond that. Note that all of the "new" code there is just the old code, although this isn't exactly like it was before #80368; it still fixes the original problem from that PR, that on RHEL 7 we were telling `iptables` to wait forever but `iptables-restore` to only wait 5 seconds. So assuming that commit is correct, that's what I'd want to merge post-1.16.0.

If you wanted something with fewer false positives but still very small, maybe we could test "if version is 1.4.21 and /etc/redhat-release exists"?

**Comment by thockin** [comment]
Does /etc/redhat-release exist in containers?  I presume not, does that
matter for you (ie. do you run kube-proxy in a container) ?

**Comment by thockin** [comment]
I don't hate that patch you linked.

**Comment by danwinship** [comment]
Yeah, `/etc/redhat-release` probably won't work reliably in containers.
Pushed #82602 with the larger commit. Let's see if it passes.

**Pull request closed without merging**